src/qibolab/instruments/tii.py

```diff
@@ -0,0 +1,239 @@
+        self.port: str
+        self.host, port = address.split(":")
+        self.port = int(port)
+
+    def connect(self):
+        """Connects to the FPGA instrument."""
+        self.is_connected = True
+
+        # Create a socket (SOCK_STREAM means a TCP socket) and send configuration
+        jsonDic = self.cfg
+        jsonDic["opCode"] = "configuration"  # opCode parameter for server
+        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
+            sock.connect((self.host, self.port))
+            sock.sendall(json.dumps(jsonDic).encode())
+
+    def setup(self, qubits, **kwargs):
+        """Configures the instrument.
+
+        A connection to the instrument needs to be established beforehand.
+        Args:
+            **kwargs: dict = A dictionary of settings loaded from the runcard:
+                kwargs['hardware_avg']
+                kwargs['repetition_duration']
+                kwargs['sampling_rate']
+                kwargs['resonator_phase']
+                kwargs['adc_trig_offset']
+                kwargs['threshold']
+                kwargs['relax_delay']
+                kwargs['max_gain']
+        Raises:
+            Exception = If attempting to set a parameter without a connection to the instrument.
+        """
+
+        if self.is_connected:
+            # Load settings
+            self.cfg = kwargs
+            jsonDic = self.cfg
+            jsonDic["opCode"] = "setup"  # opCode parameter for server
+            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
+                sock.connect((self.host, self.port))
+                sock.sendall(json.dumps(jsonDic).encode())
+
+        else:
+            raise Exception("The instrument cannot be set up, there is no connection")
+
+    def play(self, qubits, sequence, nshots, relaxation_time):
+        """Executes the sequence of instructions and retrieves the readout results.
+
+        Each readout pulse generates a separate acquisition.
+
+        Args:
+            nshots (int): parameter not used
+            relaxation_time: parameter not used
+
+        Returns:
+            Two array with real and imaginary parts if i and q (already averages)
+        """
+        # TODO clean not used parameters
+
+        jsonDic = {}
+        jsonDic["opCode"] = "execute"
+        pulsesDic = {}
+        for i, pulse in enumerate(sequence):
+            pulsesDic[str(i)] = self.convert_pulse_to_dic(pulse)
+        jsonDic["pulses"] = pulsesDic
+
+        # Create a socket and send pulse sequence to the FPGA
+        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
+            sock.connect((self.host, self.port))
+            sock.sendall(json.dumps(jsonDic).encode())
+            # read from the server a maximum of 256 bytes (enough for sequence)
+            received = sock.recv(256)
+            avg = json.loads(received)
+            avgi = avg["avgi"]
+            avgq = avg["avgq"]
+        return avgi, avgq
```

Comment:
Sorry, actually this needs to be a dictionary that maps the serial of readout pulses to `ExecutionResults`. This is because for more than one qubits you will have multiple readout pulses and each one will have its own acquisition.